File: Examples/TimeSeries/garch11.cc
// garch11 model, to illustrate using C++ for Octave bottlenecks
#include <oct.h>
#include <octave/Cell.h>

DEFUN_DLD(garch11, args, ,"garch11")
{
	// parameter of model
	ColumnVector model_params (args(0).column_vector_value());
	const double mu = model_params(0);
	const double omega = model_params(1);
	const double alpha = model_params(2);
	const double beta = model_params(3);

	// data
	const ColumnVector y (args(1).column_vector_value());

	const int n = y.rows();
	double h, hlag, e, elag, c, logdensity;
	const double pi = 3.1415926;
	int t;
	octave_value_list f_return;

	c = -log(sqrt(2.0*pi)); // constant part of log likelihood

	elag = 0.0;

	// initialize variance as variance of first 10 obs.
        hlag = 0.0;
        for (t=0; t<10;t++) hlag += (y(t)-mu)*(y(t)-mu);
        hlag = hlag/9.0;

	for (t=0; t < n; t++) { // loop over time
                e = y(t) - mu;
		h = omega + alpha*elag*elag + beta*hlag;
		logdensity += c -0.5*log(h) - 0.5*e*e/h;
		hlag = h;
		elag = e;
	}
	f_return(0) = logdensity;
	return f_return;
}

File: Examples/TimeSeries/garch11.m
% computes GARCH(1,1) log likelihood
function logL = garch11(theta, data)
    mu = theta(1,:);
    omega = theta(2,:);
    alpha = theta(3,:);
    beta = theta(4,:);
    y = data(:,1);
    e = y - mu;
    n = size(e,1);
    h = zeros(n,1);
    
    % either of these next two are reasonable choices
    h(1,:) = var(y(1:10,:));
    %h(1,:) = var(y);
    for t = 2:n
        h(t,:) = omega + alpha*e(t-1,:)^2 + beta*h(t-1,:);
    end
    logL = -log(sqrt(2*pi)) -0.5*log(h) - 0.5*(e.^2)./h;
    logL = sum(logL);
